### PR TITLE
Show preview of selected command

### PR DESCRIPTION
**Pull request by pdecat**
This addresses https://github.com/ellie/atuin/issues/405

TODO:
* [x] add an opt-in config option
* [x] unify code on one UI

### CODE REVIEW

File: src/command/client/search/interactive.rs
Comment (pdecat) [inline src/command/client/search/interactive.rs:323]: I've noticed this as soon as I added the preview chunk below the input as the cursor was place one line below the input text.

**Comment by pdecat** [comment]
Current look and feel, preview is shown below input: 
image

File: src/command/client/search/interactive.rs
Comment (pdecat) [inline src/command/client/search/interactive.rs:247]: This was initially as the following (which IMO was more readable because we are talking number of lines):
```suggestion
                    Constraint::Length(if show_help { 1 } else { 0 }),
```

But this happened to trigger the following clippy lint control:
```
warning: boolean to int conversion using if
   --> src/command/client/search/interactive.rs:247:40
    |
247 |                     Constraint::Length(if show_help { 1 } else { 0 }),
    |                                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: replace with from: `u16::from(show_help)`
    |
    = note: `show_help as u16` or `show_help.into()` can also be valid options
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#bool_to_int_with_if
    = note: `-D clippy::bool-to-int-with-if` implied by `-D warnings`
```

Just because the help height is precisely 1 line in compact mode. With a height of 2 lines, the clippy lint did not trigger.

File: src/command/client/search/interactive.rs
Comment (pdecat) [inline src/command/client/search/interactive.rs:250]: In the final version, we may want to make this configurable:
```suggestion
                    Constraint::Length(2),
```

I did not go with a dynamic height to avoid having the input position change depending on the selected result command length.

Comment (ellie) [reply src/command/client/search/interactive.rs:250]: I was just about to ask about this haha, I think it is acceptable to just have it like this regardless 🤔

Comment (pdecat) [reply src/command/client/search/interactive.rs:250]: Maybe this could be based on the longest command in history according to the current search and filter modes, ignoring the query. With a higher bound to avoid edge cases.

Comment (pdecat) [reply src/command/client/search/interactive.rs:250]: And hidden by default if no line is longer than the widget width.

Comment (pdecat) [reply src/command/client/search/interactive.rs]: Turns out it is not that perturbing to have a dynamic height based on the results even taking the query into account.

**Comment by ellie** [comment]
I think this would go nicely 🤔 Could we have some sort of limit on the max preview displayed? Nothing too much, I'd just rather avoid clobbering the UI if someone has a shell script or similar in their history

**Comment by pdecat** [comment]
> Could we have some sort of limit on the max preview displayed? Nothing too much, I'd just rather avoid clobbering the UI if someone has a shell script or similar in their history

@ellie A maximum number of lines for preview is already set to 4 here: https://github.com/ellie/atuin/pull/643/files#diff-de8948935659188c0252dc57715cff00cba21dccc39741af98b3c6a69fc42bd4R278
Maybe I should extract it into a constant to make it clearer.

File: src/command/client/search/interactive.rs
Comment (conradludgate) [inline src/command/client/search/interactive.rs:356]: this is me being grumpy about extra allocations 😅
```suggestion
        let command = if results.is_empty() {
            String::new()
        } else {
            use itertools::Itertools as _;
            let s = &results[selected].command;
            s.char_indices()
                .step_by(preview_width.into())
                .map(|(i, _)| i)
                .chain(Some(s.len()))
                .tuple_windows()
                .map(|(a, b)| &s[a..b])
                .join("\n")
        };
```

Comment (ellie) [reply src/command/client/search/interactive.rs:356]: Although the original is a bit easier to understand ;)

**Comment by pdecat** [comment]
PR converted back to draft as this is currently only implemented for compact view.

I feel like there's a lot of code duplication between the `draw` and `draw_compact` functions, which I believe isn't helping with maintenance.

Should we try and do something about that?

File: src/command/client/search/interactive.rs
Comment (conradludgate) [reply src/command/client/search/interactive.rs]: I've worked hard to make the renderer efficient!

**Comment by pdecat** [comment]
Hmm, running `atuin stats` from this branch, I'm getting the following error:

```
RUST_BACKTRACE=1 atuin stats
thread 'main' panicked at 'called `Option::unwrap()` on a `None` value', src/command/client/stats.rs:30:62
stack backtrace:
   0: rust_begin_unwind
             at /rustc/d5a82bbd26e1ad8b7401f6a718a9c57c96905483/library/std/src/panicking.rs:575:5
   1: core::panicking::panic_fmt
             at /rustc/d5a82bbd26e1ad8b7401f6a718a9c57c96905483/library/core/src/panicking.rs:64:14
   2: core::panicking::panic
             at /rustc/d5a82bbd26e1ad8b7401f6a718a9c57c96905483/library/core/src/panicking.rs:111:5
   3: core::option::Option<T>::unwrap
             at /rustc/d5a82bbd26e1ad8b7401f6a718a9c57c96905483/library/core/src/option.rs:778:21
   4: atuin::command::client::stats::compute_stats
             at ./src/command/client/stats.rs:30:20
   5: atuin::command::client::stats::Cmd::run::{{closure}}
             at ./src/command/client/stats.rs:87:9
   6: atuin::command::client::Cmd::run::{{closure}}
             at ./src/command/client.rs:55:64
   7: <core::pin::Pin<P> as core::future::future::Future>::poll
             at /rustc/d5a82bbd26e1ad8b7401f6a718a9c57c96905483/library/core/src/future/future.rs:124:9
   8: tokio::runtime::scheduler::current_thread::CoreGuard::block_on::{{closure}}::{{closure}}::{{closure}}
             at /home/patrick/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.25.0/src/runtime/scheduler/current_thread.rs:541:57
   9: tokio::runtime::coop::with_budget
             at /home/patrick/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.25.0/src/runtime/coop.rs:102:5
  10: tokio::runtime::coop::budget
             at /home/patrick/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.25.0/src/runtime/coop.rs:68:5
  11: tokio::runtime::scheduler::current_thread::CoreGuard::block_on::{{closure}}::{{closure}}
             at /home/patrick/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.25.0/src/runtime/scheduler/current_thread.rs:541:25
  12: tokio::runtime::scheduler::current_thread::Context::enter
             at /home/patrick/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.25.0/src/runtime/scheduler/current_thread.rs:350:19
  13: tokio::runtime::scheduler::current_thread::CoreGuard::block_on::{{closure}}
             at /home/patrick/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.25.0/src/runtime/scheduler/current_thread.rs:540:36
  14: tokio::runtime::scheduler::current_thread::CoreGuard::enter::{{closure}}
             at /home/patrick/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.25.0/src/runtime/scheduler/current_thread.rs:615:57
  15: tokio::macros::scoped_tls::ScopedKey<T>::set
             at /home/patrick/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.25.0/src/macros/scoped_tls.rs:61:9
  16: tokio::runtime::scheduler::current_thread::CoreGuard::enter
             at /home/patrick/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.25.0/src/runtime/scheduler/current_thread.rs:615:27
  17: tokio::runtime::scheduler::current_thread::CoreGuard::block_on
             at /home/patrick/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.25.0/src/runtime/scheduler/current_thread.rs:530:19
  18: tokio::runtime::scheduler::current_thread::CurrentThread::block_on
             at /home/patrick/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.25.0/src/runtime/scheduler/current_thread.rs:154:24
  19: tokio::runtime::runtime::Runtime::block_on
             at /home/patrick/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.25.0/src/runtime/runtime.rs:282:47
  20: atuin::command::client::Cmd::run
             at ./src/command/client.rs:52:9
  21: atuin::command::AtuinCmd::run
             at ./src/command/mod.rs:51:37
  22: atuin::Atuin::run
             at ./src/main.rs:37:9
  23: atuin::main
             at ./src/main.rs:42:5
  24: core::ops::function::FnOnce::call_once
             at /rustc/d5a82bbd26e1ad8b7401f6a718a9c57c96905483/library/core/src/ops/function.rs:507:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

Current dev which I've just rebased on does not exhibit the same. Digging...

**Comment by conradludgate** [comment]
Line 30 would be

> .entry(i.command.split_ascii_whitespace().next().unwrap())

So if the unwrap is failing, then the command is empty 👀

**Comment by pdecat** [comment]
Reproduced the issue on main.

**Comment by ellie** [comment]
> I feel like there's a lot of code duplication between the draw and draw_compact functions, which I believe isn't helping with maintenance.
>
> Should we try and do something about that?

Yup 100%, it's a bit unclear as is. I have also been considering unifying on one UI - perhaps the compact view, as the borders are probably not needed. It'd be awesome if you want to address this as part of the PR here, otherwise we can do it separately. 

I think we need to support this on both UIs atm, regardless of whether or not the duplication is resolved in this PR. I would like it to be gated behind an opt-in config option though, at least for the next release.

**Comment by pdecat** [comment]
> I have also been considering unifying on one UI - perhaps the compact view, as the borders are probably not needed

Giving it a shot...

**Comment by pdecat** [comment]
Before (compact view at the top, normal view at the bottom):

image

**Comment by pdecat** [comment]
After the merge:

image

There may be a bit of styling missing from previous normal view.

File: src/command/client/search/interactive.rs
Comment (pdecat) [inline src/command/client/search/interactive.rs:259]: I preferred to add this than turning the function into an associated one to keep things uniform.

**Comment by pdecat** [comment]
I believe this is now ready for another round of reviews.

File: src/command/client/search/interactive.rs
Comment (conradludgate) [reply src/command/client/search/interactive.rs:203]: This no longer is the case on the recent compiler IIRC. anyway, I would prefer the original here, and if the warning comes back I would accept an `allow()`

**Comment by conradludgate** [comment]
Code looks reasonable. Thanks for working on this

File: src/command/client/search/interactive.rs
Comment (pdecat) [reply src/command/client/search/interactive.rs]: Done (still requires an `allow()`).

**Comment by dbaynard** [comment]
Showing the preview below the input is a good first step. For inline functions, though, 4 lines is often not enough. I don't think any of what I suggest here should block this PR, but it's worth considering, for a moment.

1.  Other tools allow for a preview alongside the list of entries. The length would be restricted to the window height. I'd like to see this option, at some point.
2.  Alternatively (additionally?) one could make the preview window scrollable — this would involve more work than showing a full height preview.

Hopefully when this is merged you'll get feedback and can decide how to take it forward.

**Comment by conradludgate** [comment]
Yeah, I do want to explore side panels for a few ideas we have floating around (deletion and other history controls, command 'context') so I think it would be interesting

File: src/command/client/search/interactive.rs
Comment (pdecat) [inline src/command/client/search/interactive.rs:190]: Just noticed a regression following the merge of normal and compact modes, where commands with more than 2 lines, only show 2 lines, while the max should be 4 lines. This affects both modes.

Comment (pdecat) [reply src/command/client/search/interactive.rs:195]: Issue no longer reproduces.

**Comment by pdecat** [comment]
I believe this is now ready.

**Review by ellie — APPROVED** [review]
LGTM! Thank you for the feature, and thank you for cleaning up/de-duping the render code! <3

**Pull request merged**
